cmake/HPX_SetupHwloc.cmake

```diff
@@ -13,9 +13,93 @@
 # file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
 
-find_package(Hwloc)
-if(NOT Hwloc_FOUND)
-  hpx_error(
-    "Hwloc could not be found, please specify Hwloc_ROOT to point to the correct location"
+if(NOT HPX_WITH_FETCH_HWLOC)
+  find_package(Hwloc)
+  if(NOT Hwloc_FOUND)
+    hpx_error(
+      "Hwloc could not be found, please specify Hwloc_ROOT to point to the correct location"
+    )
+  endif()
+else()
+  hpx_info(
+    "HPX_WITH_FETCH_HWLOC=${HPX_WITH_FETCH_HWLOC}, Hwloc will be fetched using CMake's FetchContent"
   )
+  if(UNIX)
+    include(FetchContent)
+    fetchcontent_declare(
+      HWLoc
+      URL https://download.open-mpi.org/release/hwloc/v2.9/hwloc-2.9.3.tar.gz
+      TLS_VERIFY true
+    )
+    if(NOT HWLoc_POPULATED)
+      fetchcontent_populate(HWLoc)
+      execute_process(
+        COMMAND
+          sh -c
+          "cd ${CMAKE_BINARY_DIR}/_deps/hwloc-src && ./configure --prefix=${CMAKE_BINARY_DIR}/_deps/hwloc-installed && make -j && make install"
+      )
+    endif()
+    set(HWLOC_ROOT "${CMAKE_BINARY_DIR}/_deps/hwloc-installed")
+    set(Hwloc_INCLUDE_DIR
+        ${HWLOC_ROOT}/include
+        CACHE INTERNAL ""
+    )
+    set(Hwloc_LIBRARY
+        ${HWLOC_ROOT}/lib/libhwloc.so
+        CACHE INTERNAL ""
+    )
+  elseif("${CMAKE_GENERATOR_PLATFORM}" STREQUAL "Win64")
+    fetchcontent_declare(
+      HWLoc
+      URL https://download.open-mpi.org/release/hwloc/v2.9/hwloc-win64-build-2.9.3.zip
+      TLS_VERIFY true
+    )
+    if(NOT HWLoc_POPULATED)
+      fetchcontent_populate(HWLoc)
+    endif()
+    set(HWLOC_ROOT
+        "${CMAKE_BINARY_DIR}/_deps/hwloc-src"
+        CACHE INTERNAL ""
+    )
+    find_package(hwloc REQUIRED PATHS ${HWLOC_ROOT} NO_DEFAULT_PATH)
+    include_directories(${HWLOC_ROOT}/include)
+    link_directories(${HWLOC_ROOT}/lib)
+    set(Hwloc_INCLUDE_DIR
+        ${HWLOC_ROOT}/include
+        CACHE INTERNAL ""
+    )
+    set(Hwloc_LIBRARY
+        ${HWLOC_ROOT}/lib/libhwloc.dll.a
+        CACHE INTERNAL ""
+    )
+  else()
+    fetchcontent_declare(
+      HWLoc
+      URL https://download.open-mpi.org/release/hwloc/v2.9/hwloc-win64-build-2.9.3.zip
+      TLS_VERIFY true
+    )
+    if(NOT HWLoc_POPULATED)
+      fetchcontent_populate(HWLoc)
+    endif()
+    set(HWLOC_ROOT
+        "${CMAKE_BINARY_DIR}/_deps/hwloc-src"
+        CACHE INTERNAL ""
+    )
+    include_directories(${HWLOC_ROOT}/include)
+    link_directories(${HWLOC_ROOT}/lib)
+    set(Hwloc_INCLUDE_DIR
+        ${HWLOC_ROOT}/include
+        CACHE INTERNAL ""
+    )
+    set(Hwloc_LIBRARY
+        ${HWLOC_ROOT}/lib/libhwloc.dll.a
+        CACHE INTERNAL ""
+    )
+  endif() # End hwloc installation
+
+  add_library(Hwloc::hwloc INTERFACE IMPORTED)
+  target_include_directories(Hwloc::hwloc INTERFACE ${Hwloc_INCLUDE_DIR})
+  target_link_libraries(Hwloc::hwloc INTERFACE ${Hwloc_LIBRARY})
+  message(${Hwloc_INCLUDE_DIR})
+  message(${Hwloc_LIBRARY})
```

Comment:
Please write a verbose message of INFO type.

Reply:
You can use `hpx_info()` instead.

Reply:
Actually I was using it to debug for a linking issue I was facing
Is it ok if I just remove it?